libsql-sqlite3/src/vectordiskann.c

```diff
@@ -345,4 +403,1048 @@ void nodeBinDebug(const DiskAnnIndex *pIndex, const BlobSpot *pBlobSpot) {
+    return SQLITE_ERROR;
+  }
+  if( vectorIdxKeyNamesRender(pKey->nKeyColumns, "index_key", columnSqlNames, sizeof(columnSqlNames)) != 0 ){
+    return SQLITE_ERROR;
+  }
+  if( vectorIdxParamsPutU64(pParams, VECTOR_INDEX_TYPE_PARAM_ID, VECTOR_INDEX_TYPE_DISKANN) != 0 ){
+    return SQLITE_ERROR;
+  }
+  if( vectorIdxParamsGetU64(pParams, VECTOR_METRIC_TYPE_PARAM_ID) == 0 ){
+    if( vectorIdxParamsPutU64(pParams, VECTOR_METRIC_TYPE_PARAM_ID, VECTOR_METRIC_TYPE_COS) != 0 ){
+      return SQLITE_ERROR;
+    }
+  }
+  if( vectorIdxParamsGetU64(pParams, VECTOR_BLOCK_SIZE_PARAM_ID) == 0 ){
+    if( vectorIdxParamsPutU64(pParams, VECTOR_BLOCK_SIZE_PARAM_ID, VECTOR_BLOCK_SIZE_DEFAULT) != 0 ){
+      return SQLITE_ERROR;
+    }
+  }
+  if( vectorIdxParamsGetF64(pParams, VECTOR_PRUNING_ALPHA_PARAM_ID) == 0 ){
+    if( vectorIdxParamsPutF64(pParams, VECTOR_PRUNING_ALPHA_PARAM_ID, VECTOR_PRUNING_ALPHA_DEFAULT) != 0 ){
+      return SQLITE_ERROR;
+    }
+  }
+  if( vectorIdxParamsGetU64(pParams, VECTOR_INSERT_L_PARAM_ID) == 0 ){
+    if( vectorIdxParamsPutU64(pParams, VECTOR_INSERT_L_PARAM_ID, VECTOR_INSERT_L_DEFAULT) != 0 ){
+      return SQLITE_ERROR;
+    }
+  }
+  if( vectorIdxParamsGetU64(pParams, VECTOR_SEARCH_L_PARAM_ID) == 0 ){
+    if( vectorIdxParamsPutU64(pParams, VECTOR_SEARCH_L_PARAM_ID, VECTOR_SEARCH_L_DEFAULT) != 0 ){
+      return SQLITE_ERROR;
+    }
+  }
+  zSql = sqlite3MPrintf(
+      db,
+      "CREATE TABLE IF NOT EXISTS %s_shadow (%s, data BLOB, PRIMARY KEY (%s))",
+      zIdxName,
+      columnSqlDefs,
+      columnSqlNames
+  );
+  rc = sqlite3_exec(db, zSql, 0, 0, 0);
+  sqlite3DbFree(db, zSql);
+  return rc;
+}
+
+int diskAnnClearIndex(sqlite3 *db, const char *zIdxName) {
+  char *zSql = sqlite3MPrintf(db, "DELETE FROM %s_shadow", zIdxName);
+  int rc = sqlite3_exec(db, zSql, 0, 0, 0);
+  sqlite3DbFree(db, zSql);
+  return rc;
+}
+
+int diskAnnDropIndex(sqlite3 *db, const char *zIdxName){
+  char *zSql = sqlite3MPrintf(db, "DROP TABLE %s_shadow", zIdxName);
+  int rc = sqlite3_exec(db, zSql, 0, 0, 0);
+  sqlite3DbFree(db, zSql);
+  return rc;
+}
+
+/*
+ * Select random row from the shadow table and set its rowid to pRowid
+ * returns SQLITE_DONE if no row found (this will be used to determine case when table is empty)
+ * TODO: we need to make this selection procedure faster - now it works in linear time
+*/
+static int diskAnnSelectRandomShadowRow(const DiskAnnIndex *pIndex, u64 *pRowid){
+  int rc;
+  sqlite3_stmt *pStmt = NULL;
+  char *zSql = NULL;
+
+  zSql = sqlite3MPrintf(pIndex->db, "SELECT rowid FROM %s ORDER BY RANDOM() LIMIT 1", pIndex->zShadow);
+  if( zSql == NULL ){
+    rc = SQLITE_NOMEM_BKPT;
+    goto out;
+  }
+  rc = sqlite3_prepare_v2(pIndex->db, zSql, -1, &pStmt, 0);
+  if( rc != SQLITE_OK ){
+    goto out;
+  }
+  rc = sqlite3_step(pStmt);
+  if( rc != SQLITE_ROW ){
+    goto out;
+  }
+  *pRowid = sqlite3_column_int64(pStmt, 0);
```

Comment:
Don't we have to check:
```
  if( sqlite3_column_type(pStmt, 0) != SQLITE_INTEGER ){
    rc = SQLITE_ERROR;
    goto out;
  }
```
like in the function below?

Reply:
I added asserts in similar places everywhere (since we have full control over the query - we actually can just assert this invariant, I think)